### PR TITLE
Feature | Postgres Support

### PR DESCRIPTION
**Pull request by ksimenic**
This PR merges 2 existing PRs: #60 and #63 and potentially solves https://github.com/Sammyjo20/laravel-haystack/issues/61

Additionally, both PRs are standardised (pint) and upgraded with following changes:
- 60 - improved description in config file
- 60 - improved env vars fallbacks and added new env var as override
- 60 - added missing return types on models
- 60 - improve TestCase logic, add return types and fix failing test
- 63 - extracted `DB::connection instanceof` check to separate function and take db_connection configuration var (from 60) into account

### CODE REVIEW

**Comment by ksimenic** [comment]
Additional comment for testing. Since sqlite is used for testing, it is not that straight forward to implement tests for mysql/pgsql. That would also require additional work for both local development and github actions. However, if adding tests is necessary I can work on them in the scope of this PR or in some other PR.

**Comment by Sammyjo20** [comment]
Many thanks for this PR I will have a review now. If you know how to build a testing matrix with SQLite, MySQL and Postgres that would be fantastic. Might that be something you could add to this PR please?

File: src/Helpers/SerializationHelper.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    /**
+     * Serialize the given value.
+     */
+    public static function serialize(mixed $value): string
+    {
+        $serialized = serialize($value);
+
+        return self::isPgsql()
+            ? base64_encode($serialized)
+            : $serialized;
```
Comment (Sammyjo20) [inline src/Helpers/SerializationHelper.php:22]: This doesn't need to be put on two lines

```suggestion
        return self::isPgsql() ? base64_encode($serialized) : $serialized;
```

File: src/Helpers/SerializationHelper.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+     * Unserialize the given value.
+     */
+    public static function unserialize(
+        string $serialized,
+        array $options = []
+    ): mixed {
+        if (
+            self::isPgsql()
+            && ! Str::contains($serialized, [':', ';'])
+        ) {
```
Comment (Sammyjo20) [inline src/Helpers/SerializationHelper.php:35]:
```suggestion
        if (self::isPgsql() && ! Str::contains($serialized, [':', ';'])) {
```

File: src/Helpers/SerializationHelper.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            : $serialized;
+    }
+
+    /**
+     * Unserialize the given value.
+     */
+    public static function unserialize(
+        string $serialized,
+        array $options = []
+    ): mixed {
```
Comment (Sammyjo20) [inline src/Helpers/SerializationHelper.php:31]:
```suggestion
    public static function unserialize(string $serialized, array $options = []): mixed 
    {
```

**Review by Sammyjo20 — CHANGES_REQUESTED** [review]
Thank you for this

File: src/Helpers/SerializationHelper.php (hunk 1 above)
Comment (ksimenic) [reply src/Helpers/SerializationHelper.php]: Fixed.

File: src/Helpers/SerializationHelper.php (hunk 2 above)
Comment (ksimenic) [reply src/Helpers/SerializationHelper.php]: Fixed.

File: src/Helpers/SerializationHelper.php (hunk 3 above)
Comment (ksimenic) [reply src/Helpers/SerializationHelper.php]: Fixed.

**Comment by ksimenic** [comment]
@Sammyjo20 

Added tests, as requested. Additionally, I ran pint on whole project (not just on updated files) and fixed code style per your suggestion.

There are 2 things you should be aware of.

1. Running tests locally
Since now we have tests on MySQL and PgSQL, to run the tests locally we need both MySQL and PgSQL server. 
I am using DBnging locally. Both MySQL and PgSQL are created with default values except the port. For MySQL port is 33066 and for PgSQL port is 54321. Additionally, once server is created and started you must create database named "testing" on both servers.
image

2. Running tests in matrix
Needed to drop support for "windows-latest" since you can't create "services" (postgres, mysql) there.
image

**Comment by ksimenic** [comment]
Hey @Sammyjo20. I am just wondering if there is a plan to release this any time soon? If not, I will need to unblock my projects by using my own fork. No problem at all if there is no plan to release it. Thank you!

**Comment by Sammyjo20** [comment]
Hey @ksimenic I am hoping to release this at the weekend if this works for you?

**Comment by ksimenic** [comment]
@Sammyjo20 works for me. Thank you!

**Review by Sammyjo20 — APPROVED** [review]
Happy with this 💪

**Comment by Sammyjo20** [comment]
Thank you @ksimenic for your help on this!

**Comment by ksimenic** [comment]
Hey @Sammyjo20. Thank you for releasing this! Just wondering why you dropped support for Laravel 8.x?

**Comment by Sammyjo20** [comment]
Hey @ksimenic I've dropped support for Laravel 8 because it's no longer getting any security updates or officially supported. 

image

**Pull request merged**